src/compiler/checker.ts

```diff
@@ -7571,6 +7574,78 @@ namespace ts {
             return undefined;
         }
 
+        function shouldAcquireLiteralType(literalNode: StringLiteral) {

```

Comment:
An alternative way to structure this is to have an `isEqualityComparisonOperand` function that returns true if the node is the operand of an equality comparison (including switch and case). You could then add a call to this function to the existing code in `checkStringLiteralExpression`. As it is now, this function mixes two concerns (equality comparison operand and presence of string literal contextual type) that really are separate.